lib/floe/workflow/runner/docker.rb

```diff
@@ -45,10 +46,66 @@ def cleanup(runner_context)
           delete_secret(secrets_file)    if secrets_file
         end
 
+        def wait(timeout: nil, events: %i[create update delete], &block)
+          until_timestamp = Time.now.utc + timeout if timeout
+
+          r, w = IO.pipe
+
+          pid = AwesomeSpawn.run_detached(
+            self.class::DOCKER_COMMAND, :err => :out, :out => w, :params => wait_params(until_timestamp)
+          )
+
+          w.close
+
+          loop do
+            readable_timeout = until_timestamp - Time.now.utc if until_timestamp
+
+            # Wait for our end of the pipe to be readable and if it didn't timeout
+            # get the events from stdout
+            next if r.wait_readable(readable_timeout).nil?
+
+            # Get all events while the pipe is readable
+            notices = []
+            while r.ready?
+              notice = r.gets
+
+              # If the process has exited `r.gets` returns `nil` and the pipe is
+              # always `ready?`
+              break if notice.nil?
+
+              ref, event, state = parse_notice(notice)
+              next unless events.include?(event)
+
+              state ||= inspect_container(ref)&.dig("State")
+
+              runner_context = {"container_ref" => ref, "container_state" => state}
+
+              notices << runner_context
+            end
+
+            # If we're given a block yield the events otherwise return them
+            if block
+              notices.each(&block)
+            else
+              # Terminate the `docker events` process before returning the events
+              Process.kill("TERM", pid) rescue Errno::ESRCH
+              return notices
+            end
+
+            # Check that the `docker events` process is still alive
+            Process.kill(0, pid)
+          rescue Errno::ESRCH
+            # Break out of the loop if the `docker events` process has exited
+            break
+          end
+        ensure
+          r.close
```

Comment:
Does this need a rescue/check in case the pipe is closed prematurely or is that not possible? (I'm thinking if there's an error and the pipe is broken).  Maybe something like the following (or maybe close just handles it already being closed and this all doesn't matter)

```suggestion
          r.close unless r.closed?
```

Reply:
> or maybe close just handles it already being closed and this all doesn't matter

This, if it were somehow closed already `r.close` doesn't raise an exception
```
>> r, w = IO.pipe
=> [#<IO:fd 5>, #<IO:fd 6>]
>> r.close
=> nil
>> r.close
=> nil
>> r.close
=> nil
```